Add SRGAN and datamodules for super resolution

chris-clem opened the pull request:
## What does this PR do?

Adds a SRGAN implementation to bolts as proposed in #412.

Closes #412 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?
- [x] Add train logs and example images

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @chris-clem! Thanks for opening this PR. 











* In the file [`pl_bolts/models/gans/srgan/components.py`](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/01981c29a35481a7063bd62cfc16145d52e1cebc/pl_bolts/models/gans/srgan/components.py):

> [Line 1:121](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/01981c29a35481a7063bd62cfc16145d52e1cebc/pl_bolts/models/gans/srgan/components.py#L1): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (158 > 120 characters)





Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

chris-clem commented:
Should I also upload the `lightning_logs `containing the checkpoints?

Borda reviewed (changes requested):
Pls limit change just to adding a new module (your automatic formating is changing some other unrelated files) and do not add any trained model, thank you for your understanding and making the review easier 🐰

oke-aditya commented on pl_bolts/models/gans/srgan/components.py:
```diff
@@ -0,0 +1,150 @@
+    warn_missing_pkg("torchvision")  # pragma: no-cover
+
+
+class ResidualBlock(nn.Module):
```
I think Residual Block is far too often used in our codebase (SSL too) 
Marking it for #427

akihironitta replied:
@oke-aditya Thanks for marking it :]

oke-aditya commented on pl_bolts/models/gans/srgan/components.py:
```diff
@@ -0,0 +1,150 @@
+        return x + self.block(x)
+
+
+class SRGANGenerator(nn.Module):
```
I think we can do something as I mentioned in DCGAN #403 . Marking again for discussion in #427

oke-aditya commented on pl_bolts/models/gans/srgan/components.py:
```diff
@@ -0,0 +1,150 @@
+        return x
+
+
+class VGG19FeatureExtractor(nn.Module):
```
I think FeatureExtractor using this method is also common.
I think some part of this code matches with #475

oke-aditya reviewed (commented):
I have few thoughts, I think we can have very nice support for GANs.
Like [mimicry](https://github.com/kwotsin/mimicry). If we can decompose the generators and discriminators, a more generic use might be possible.

chris-clem commented:
> Pls limit change just to adding a new module (your automatic formating is changing some other unrelated files) and do not add any trained model, thank you for your understanding and making the review easier 🐰

@Borda, sorry, I updated the PR to revert black changes and removed the checkpoints.

akihironitta commented:
@chris-clem Would you mind resolving the conflicts?

chris-clem commented:
> @chris-clem Would you mind resolving the conflicts? (Also, I might need some time to review this PR as I'm not familiar with this model. I'll try to check the original paper and some other implementations as soon as possible!)

@akihironitta Done! The tests fail for Python 3.6 on ubuntu and mac. Do you know why? Has it something to do with #471?

ananyahjha93 commented:
@chris-clem yes its because of #471. Let me take a look

ananyahjha93 commented:
@chris-clem rebase and resolve conflicts, good to go!

akihironitta commented on pl_bolts/datamodules/sr_datamodule.py:
```diff
@@ -0,0 +1,72 @@
+            batch_size: How many samples per batch to load
+            num_workers: How many workers to use for loading data
+            shuffle: If true shuffles the train data every epoch
+            pin_memory: If true, the data loader will copy Tensors into CUDA pinned memory before
+                        returning them
```
```suggestion
            pin_memory: If true, the data loader will copy Tensors into CUDA pinned memory before
                returning them
```

akihironitta commented on tests/models/test_scripts.py:
```diff
@@ -19,6 +19,34 @@ def test_cli_run_basic_gan(cli_args):
+@pytest.mark.parametrize(
+    "cli_args",
+    [
+        f"--dataset mnist --scale_factor 4 --data_dir {DATASETS_PATH} --max_epochs 1"
+        " --batch_size 2 --limit_train_batches 2 --limit_val_batches 2"
```
How about  adding `--fast_dev_run` for faster testing?
```suggestion
        f"--dataset mnist --scale_factor 4 --data_dir {DATASETS_PATH} --max_epochs 1"
        " --batch_size 2 --limit_train_batches 2 --limit_val_batches 2 --fast_dev_run 1"
```

akihironitta commented on tests/models/test_scripts.py:
```diff
@@ -19,6 +19,34 @@ def test_cli_run_basic_gan(cli_args):
+@pytest.mark.parametrize(
+    "cli_args",
+    [
+        f"--dataset mnist --data_dir {DATASETS_PATH} --max_epochs 1"
+        " --batch_size 2 --limit_train_batches 2 --limit_val_batches 2"
```
Same as above.
```suggestion
        f"--dataset mnist --data_dir {DATASETS_PATH} --max_epochs 1"
        " --batch_size 2 --limit_train_batches 2 --limit_val_batches 2 --fast_dev_run 1"
```

akihironitta reviewed (commented):
Haven't checked the logic yet, but I left some general comments below. Also, the tests you added are failing... Would you mind having a look at them, too?

akihironitta commented:
@chris-clem I believe the change in https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/466/commits/c1d236ac4fb92622a56fb0d5b8b5084b49ae251d is making the tests fail on Python 3.6 as reported in #471, so could you revert it?

chris-clem commented:
> @chris-clem I believe the change in [c1d236a](https://github.com/PyTorchLightning/pytorch-lightning-bolts/commit/c1d236ac4fb92622a56fb0d5b8b5084b49ae251d) is making the tests fail on Python 3.6 as reported in #471, so could you revert it?

Ah right. I reverted in 9ad817c, but two tests still fail with `TypeError: 'NoneType' object is not iterable`. I removed `--fast_dev_run` completely in f11b4ae, but the same tests still fail. I am not sure what the issue is now with the tests. For pytest (macOS-10.15, 3.6, latest) and pytest (macOS-10.15, 3.6, latest) `test_cli_run_srgan` and `test_cli_run_srresnet` fail with the following error:

```python
=================================== FAILURES ===================================
_ test_cli_run_srgan[--dataset mnist --scale_factor 4 --data_dir /home/runner/work/pytorch-lightning-bolts/pytorch-lightning-bolts/datasets --max_epochs 1 --batch_size 2 --limit_train_batches 2 --limit_val_batches 2] _

cli_args = '--dataset mnist --scale_factor 4 --data_dir /home/runner/work/pytorch-lightning-bolts/pytorch-lightning-bolts/datasets --max_epochs 1 --batch_size 2 --limit_train_batches 2 --limit_val_batches 2'

    @pytest.mark.parametrize(
        "cli_args",
        [
            f"--dataset mnist --scale_factor 4 --data_dir {DATASETS_PATH} --max_epochs 1"
            " --batch_size 2 --limit_train_batches 2 --limit_val_batches 2"
        ],
    )
    def test_cli_run_srgan(cli_args):
        from pl_bolts.models.gans.srgan.srgan_module import cli_main
    
        with mock.patch("argparse._sys.argv", ["any.py"] + cli_args.strip().split()):
>           cli_main()

tests/models/test_scripts.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pl_bolts/models/gans/srgan/srgan_module.py:181: in cli_main
    args = parse_args(args, pl_module_cls)
pl_bolts/models/gans/srgan/utils.py:22: in parse_args
    parser = SRDataModule.add_argparse_args(parser)
/opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/pytorch_lightning/core/datamodule.py:282: in add_argparse_args
    for at in cls.get_init_arguments_and_types()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pl_bolts.datamodules.sr_datamodule.SRDataModule'>

    @classmethod
    def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
        r"""Scans the DataModule signature and returns argument names, types and default values.
        Returns:
            List with tuples of 3 values:
            (argument name, set with argument types, argument default value).
        """
        datamodule_default_params = inspect.signature(cls.__init__).parameters
        name_type_default = []
        for arg in datamodule_default_params:
            arg_type = datamodule_default_params[arg].annotation
            arg_default = datamodule_default_params[arg].default
            try:
>               arg_types = tuple(arg_type.__args__)
E               TypeError: 'NoneType' object is not iterable

/opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/pytorch_lightning/core/datamodule.py:362: TypeError
```

`test_cli_run_basic_gan` passes:

```python
@pytest.mark.parametrize('cli_args', [
    f'--dataset mnist --data_dir {DATASETS_PATH} --max_epochs 1'
    ' --batch_size 2 --limit_train_batches 2 --limit_val_batches 2',
    f'--dataset cifar10 --data_dir {DATASETS_PATH} --max_epochs 1'
    ' --batch_size 2 --limit_train_batches 2 --limit_val_batches 2',
])
def test_cli_run_basic_gan(cli_args):
    from pl_bolts.models.gans.basic.basic_gan_module import cli_main

    with mock.patch("argparse._sys.argv", ["any.py"] + cli_args.strip().split()):
        cli_main()
```

`test_cli_run_srgan` fails:

```python
@pytest.mark.parametrize(
    "cli_args",
    [
        f"--dataset mnist --scale_factor 4 --data_dir {DATASETS_PATH} --max_epochs 1"
        " --batch_size 2 --limit_train_batches 2 --limit_val_batches 2"
    ],
)
def test_cli_run_srgan(cli_args):
    from pl_bolts.models.gans.srgan.srgan_module import cli_main

    with mock.patch("argparse._sys.argv", ["any.py"] + cli_args.strip().split()):
        cli_main()
```

@akihironitta, do you see what could cause the `TypeError: 'NoneType' object is not iterable` error?

Maybe because `SRDataModule` has some args without default values that cause the error in `get_init_arguments_and_types`?

```python
class SRDataModule(LightningDataModule):
    """
    DataModule for Super Resolution datasets.

    Example::
        from pl_bolts.datamodules import SRDataModule
        from pl_bolts.datasets.mnist_dataset import SRMNISTDataset

        dataset_dev = SRMNISTDataset(scale_factor=4, root=".", train=True)
        dataset_train, dataset_val = random_split(dataset_dev, lengths=[55_000, 5_000])
        dataset_test = SRMNISTDataset(scale_factor=4, root=".", train=True)
        dm = SRDataModule(dataset_train, dataset_val, dataset_test)
    """

    def __init__(
        self,
        dataset_train: Dataset,
        dataset_val: Dataset,
        dataset_test: Dataset,
        batch_size: int = 16,
        shuffle: bool = True,
        num_workers: int = 8,
        pin_memory: bool = True,
        drop_last: bool = True,
        *args,
        **kwargs,
    ) -> None:
```

chris-clem commented:
> @chris-clem seems like all issues were resolved, correct?

@Borda The failing tests I mentioned in the [comment ](https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/466#issuecomment-757478475) above are still failing. I will look into it again during the week.

akihironitta commented on pl_bolts/datamodules/sr_datamodule.py:
```diff
@@ -0,0 +1,72 @@
+from torch.utils.data import DataLoader, Dataset
+
+
+class SRDataModule(LightningDataModule):
```
It seems to me that this class isn't really "SR" thing, is it...?

chris-clem replied:
Haha yes indeed, that is a good observation. What would you suggest I should do with it? 

Is it better to add the dataset preparation ([`prepare_datasets()` in `srgan/utils.py`](https://github.com/chris-clem/pytorch-lightning-bolts/blob/18132e781b42b5c0edd7daffe8cef05437b5ee96/pl_bolts/models/gans/srgan/utils.py#L30-L51)) into the DataModule class?

chris-clem replied:
@akihironitta The problem with `VisionDataModule` is that it only works with `torchvision` datasets that have a boolean `train` flag. So it would work for `MNIST`, but not for `STL10` or `CelebA` (they have a `split` flag).

I could update `VisionDataModule` first so that it works with all `torchvision` datasets and then use it here?

chris-clem replied:
Yes I can do that. At the moment, I have a bit less time then when I created these PRs but that change should be not too complicated. I try to do it next weekend.

akihironitta commented on pl_bolts/datasets/mnist_dataset.py:
```diff
@@ -29,7 +32,7 @@ def __getitem__(self, idx):
         # doing this so that it is consistent with all other datasets
         # to return a PIL Image
-        img = Image.fromarray(img.numpy(), mode='L')
+        img = PIL.Image.fromarray(img.numpy(), mode="L")
```
```suggestion
        img = Image.fromarray(img.numpy(), mode="L")
```

akihironitta commented on pl_bolts/datasets/mnist_dataset.py:
```diff
@@ -42,3 +45,29 @@ def __getitem__(self, idx):
+        self.image_channels = 1
+
+    def _get_image(self, index: int):
+        return PIL.Image.fromarray(self.data[index].numpy(), mode="L")
```
```suggestion
        return Image.fromarray(self.data[index].numpy(), mode="L")
```

akihironitta commented on pl_bolts/datasets/sr_celeba_dataset.py:
```diff
@@ -0,0 +1,34 @@
+from pl_bolts.utils.warnings import warn_missing_pkg
+
+if _PIL_AVAILABLE:
+    import PIL
```
Not a strong opinion, but I prefer importing only the item needed in this module...
```suggestion
    from PIL import Image
```

chris-clem commented on pl_bolts/callbacks/vision/sr_image_logger.py:
```diff
@@ -0,0 +1,47 @@
+    warn_missing_pkg("torchvision")  # pragma: no-cover
+
+
+class SRImageLoggerCallback(Callback):
```
Good point, done!

akihironitta commented on docs/source/gans.rst:
```diff
@@ -42,6 +42,7 @@ Loss curves:
 .. autoclass:: pl_bolts.models.gans.GAN
    :noindex:
 
+
```
```suggestion
```

akihironitta commented on docs/source/gans.rst:
```diff
@@ -50,7 +51,7 @@ PyTorch's `examples <https://github.com/pytorch/examples/blob/master/dcgan/main.
 Implemented by:
 
-    - `Christoph Clement <https://github.com/chris-clem>`_
+    - `Christoph Clement <https://github.com/chris-clem>`_   
```
```suggestion
    - `Christoph Clement <https://github.com/chris-clem>`_
```

akihironitta commented on pl_bolts/datasets/mnist_dataset.py:
```diff
@@ -29,7 +32,7 @@ def __getitem__(self, idx):
         # doing this so that it is consistent with all other datasets
         # to return a PIL Image
-        img = Image.fromarray(img.numpy(), mode='L')
+        img = Image.fromarray(img.numpy(), mode="L")
```
```suggestion
        img = Image.fromarray(img.numpy(), mode='L')
```

akihironitta reviewed (commented):
@chris-clem It'd be fantastic to have this model in Bolts! Let me know your thoughts on my suggestions below.

akihironitta commented on pl_bolts/datasets/sr_celeba_dataset.py:
```diff
@@ -0,0 +1,34 @@
+
+if _PIL_AVAILABLE:
+    from PIL import Image
+else:
+    warn_missing_pkg("PIL", pypi_name="Pillow")  # pragma: no-cover
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("PIL", pypi_name="Pillow")
```

akihironitta commented on pl_bolts/datasets/sr_celeba_dataset.py:
```diff
@@ -0,0 +1,34 @@
+
+if _TORCHVISION_AVAILABLE:
+    from torchvision.datasets import CelebA
+else:
+    warn_missing_pkg("torchvision")  # pragma: no-cover
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("torchvision")
```

akihironitta commented on pl_bolts/datasets/sr_dataset_mixin.py:
```diff
@@ -0,0 +1,53 @@
+
+if _PIL_AVAILABLE:
+    from PIL import Image
+else:
+    warn_missing_pkg("PIL", pypi_name="Pillow")  # pragma: no-cover
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("PIL", pypi_name="Pillow")
```

akihironitta commented on pl_bolts/datasets/sr_dataset_mixin.py:
```diff
@@ -0,0 +1,53 @@
+
+if _TORCHVISION_AVAILABLE:
+    from torchvision import transforms as transform_lib
+else:
+    warn_missing_pkg("torchvision")  # pragma: no-cover
+
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("torchvision")

```

akihironitta commented on pl_bolts/datasets/sr_stl10_dataset.py:
```diff
@@ -0,0 +1,34 @@
+
+if _PIL_AVAILABLE:
+    import PIL
+else:
+    warn_missing_pkg("PIL", pypi_name="Pillow")  # pragma: no-cover
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("PIL", pypi_name="Pillow")
```

akihironitta commented on pl_bolts/datasets/sr_stl10_dataset.py:
```diff
@@ -0,0 +1,34 @@
+
+if _TORCHVISION_AVAILABLE:
+    from torchvision.datasets import STL10
+else:
+    warn_missing_pkg("torchvision")  # pragma: no-cover
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("torchvision")
```

akihironitta commented on pl_bolts/models/gans/srgan/components.py:
```diff
@@ -0,0 +1,154 @@
+
+if _TORCHVISION_AVAILABLE:
+    from torchvision.models import vgg19
+else:
+    warn_missing_pkg("torchvision")  # pragma: no-cover
```
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("torchvision")
```

Borda commented:
@akihironitta is this ready?
@chris-clem mind resolve conflicts? :]

akihironitta reviewed (commented):
@chris-clem Sorry for the extreme delay. Your work is great! I added some commits to your branch directly, so would you mind having a look at my changes? If there's something I changed wrongly, please let me know. I checked the paper and the repo you referenced in the code, and I think the model implementation is good to go once the failing tests are resolved!

All my concern is, however, about the datasets/datamodule which I commented in https://github.com/PyTorchLightning/lightning-bolts/pull/466/files#r559892626.

@PyTorchLightning/core-bolts We could merge this PR first and refactor the datasets/datamodule later in another PR, but what do you think?

chris-clem commented on tests/models/test_scripts.py:
```diff
@@ -27,6 +27,54 @@ def test_cli_run_dcgan(cli_args):
+        cli_main()
+
+
+# TODO: this test is hanging (runs for more then 10min) so we need to use GPU or optimize it...
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
+@pytest.mark.parametrize(
+    'cli_args', [
+        f' --data_dir {DATASETS_PATH}'
+        ' --max_epochs 1'
+        ' --limit_train_batches 2'
+        ' --limit_val_batches 2'
+        ' --batch_size 2'
+        ' --encoder resnet18'
+    ]
+)
+def test_cli_run_cpc(cli_args):
+    from pl_bolts.models.self_supervised.cpc.cpc_module import cli_main
+
+    cli_args = cli_args.strip().split(' ') if cli_args else []
+    with mock.patch("argparse._sys.argv", ["any.py"] + cli_args):
+        cli_main()
+
+
```
Should I remove it in this PR?

Borda commented:
@chris-clem @akihironitta do you think we can get this done any time soon? :]

chris-clem commented:
> @chris-clem @akihironitta do you think we can get this done any time soon? :]

Yess! @akihironitta how should we tackle it?

akihironitta reviewed (commented):
It basically looks good to me :]

Only two possible improvements on code structure are:
- [ ] To define `prepare_datasets()` in `pl_bolts/datasets/*.py` rather than in `pl_bolts/models/gans/srgan/utils.py`
- [ ] To rename `SRDataModule` to something else because the class isn't bounded only to super resolution domain.

Once the above are resolved, it all looks good to me.

akihironitta commented on pl_bolts/datasets/sr_dataset_mixin.py:
```diff
@@ -0,0 +1,53 @@
+        ])
+
+    @property
+    def lr_transforms(self):
+        # Scale range of LR images to [0, 1]
+        return transform_lib.Compose([
```
Same applies to `hr_transforms`.

chris-clem commented:
> It basically looks good to me :]
> 
> Only two possible improvements on code structure are:
> 
> * [ ]  To define `prepare_datasets()` in `pl_bolts/datasets/*.py` rather than in `pl_bolts/models/gans/srgan/utils.py`
> * [ ]  To rename `SRDataModule` to something else because the class isn't bounded only to super resolution domain.
> 
> Once the above are resolved, it all looks good to me.

Hi @akihironitta,

thank you for your suggestions. I implemented them.

chris-clem commented on pl_bolts/datasets/sr_dataset_mixin.py:
```diff
@@ -0,0 +1,53 @@
+            transform_lib.Normalize(mean=(-1.0, ) * self.image_channels, std=(2.0, ) * self.image_channels),
+            transform_lib.ToPILImage(),
+            transform_lib.Resize(self.lr_image_size, Image.BICUBIC),
+            transform_lib.ToTensor(),
```
If I remove the conversion, the tests fail:

```
E         File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/torchvision/transforms/transforms.py", line 244, in __call__
E           return F.resize(img, self.size, self.interpolation)
E         File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/torchvision/transforms/functional.py", line 319, in resize
E           raise TypeError('img should be PIL Image. Got {}'.format(type(img)))
E       TypeError: img should be PIL Image. Got <class 'torch.Tensor'>
```

chris-clem commented:
> @chris-clem how is it going here, ready for review? :rabbit: 

@Borda Yes I think so!

akihironitta commented on CHANGELOG.md:
```diff
@@ -8,6 +8,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Added `SRGAN` model ([#466](https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/466))
+
+- Added `EMNISTDataModule`, `BinaryEMNISTDataModule`, and `BinaryEMNIST` dataset ([#676](https://github.com/PyTorchLightning/lightning-bolts/pull/676))
```
```suggestion
- Added `SRGAN` model ([#466](https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/466))

- Added `EMNISTDataModule`, `BinaryEMNISTDataModule`, and `BinaryEMNIST` dataset ([#676](https://github.com/PyTorchLightning/lightning-bolts/pull/676))

- Added `SRImageLoggerCallback` ([#676](https://github.com/PyTorchLightning/lightning-bolts/pull/676))
```

akihironitta commented:
@chris-clem Hi, thank you again for all your work and patience for nearly a year :sweat_smile: I'll try to be active on this PR for the next couple of days until we have this PR merged to master!

akihironitta reviewed (commented):
@chris-clem I was thinking about the datamodule again, and we could probably add `VisionDataModule.from_datasets(train_dataset, val_dataset, test_dataset)` to `VisionDataModule` instead of introducing the new datamodule class `TVTDataModule`, but I think it can be addressed/refactored in another PR.

It seems like the added tests are failing. Once the failing tests are fixed, I believe we can land this PR. It'd be great if you could push fixes, but I'll look into them anyway soon (hopefully) :]

chris-clem commented:
> @chris-clem I was thinking about the datamodule again, and we could probably add a classmethod `VisionDataModule.from_datasets(train_dataset, val_dataset, test_dataset)` to `VisionDataModule` (which we have in Flash https://lightning-flash.readthedocs.io/en/latest/api/generated/flash.core.data.data_module.DataModule.html#flash.core.data.data_module.DataModule.from_datasets) instead of introducing the new datamodule class `TVTDataModule`, but I think it can be addressed/refactored in another PR.
> 
> It seems like the added tests are failing. Once the failing tests are fixed, I believe we can land this PR. It'd be great if you could push fixes, but I'll look into them anyway soon (hopefully) :]

Hi @akihironitta, that idea makes a lot of sense! Should I implement it in this PR?

Regarding the failing test: the only fail when Python 3.6, latest is used. This is the trace:

```
_ test_cli_run_srgan[--dataset mnist --scale_factor 4 --data_dir /Users/runner/work/lightning-bolts/lightning-bolts/datasets --max_epochs 1 --max_steps 2 --batch_size 4] _

cli_args = '--dataset mnist --scale_factor 4 --data_dir /Users/runner/work/lightning-bolts/lightning-bolts/datasets --max_epochs 1 --max_steps 2 --batch_size 4'

    @pytest.mark.parametrize("cli_args", ["--dataset mnist --scale_factor 4" + _DEFAULT_ARGS])
    def test_cli_run_srgan(cli_args):
        from pl_bolts.models.gans.srgan.srgan_module import cli_main
    
        with mock.patch("argparse._sys.argv", ["any.py"] + cli_args.strip().split()):
>           cli_main()

tests/models/test_scripts.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pl_bolts/models/gans/srgan/srgan_module.py:194: in cli_main
    parser = TVTDataModule.add_argparse_args(parser)
../../../hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pytorch_lightning/core/datamodule.py:285: in add_argparse_args
    return add_argparse_args(cls, parent_parser, **kwargs)
../../../hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pytorch_lightning/utilities/argparse.py:202: in add_argparse_args
    args_and_types = get_init_arguments_and_types(symbol)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <function TVTDataModule.__init__ at 0x152c96ae8>

    def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
        r"""Scans the class signature and returns argument names, types and default values.
    
        Returns:
            List with tuples of 3 values:
            (argument name, set with argument types, argument default value).
    
        Examples:
    
            >>> from pytorch_lightning import Trainer
            >>> args = get_init_arguments_and_types(Trainer)
    
        """
        cls_default_params = inspect.signature(cls).parameters
        name_type_default = []
        for arg in cls_default_params:
            arg_type = cls_default_params[arg].annotation
            arg_default = cls_default_params[arg].default
            try:
>               arg_types = tuple(arg_type.__args__)
E               TypeError: 'NoneType' object is not iterable

../../../hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/pytorch_lightning/utilities/argparse.py:128: TypeError
```

I do not really get the problem, do you?

akihironitta commented:
> Hi @akihironitta, that idea makes a lot of sense! Should I implement it in this PR?

Let's land this PR first and do that later in another PR if other core members agree :)

> Regarding the failing test: the only fail when Python 3.6, latest is used. This is the trace:
> ...
> I do not really get the problem, do you?

I don't either :sweat_smile: I'll give it a try to fix it now :)

akihironitta commented:
@chris-clem You kindly reported the failing tests ages ago in https://github.com/PyTorchLightning/lightning-bolts/pull/466#issuecomment-757478475, but I somehow missed it... Sorry (still trying debugging)

akihironitta commented:
### TODO
- [ ] Enable the GPU testing (which is temporarily disabled at 40d3770)

akihironitta commented:
Just for the note, I'm sure this error comes from PL particularly when used with Python 3.6.

<details>
<summary>A Minimal script for reproduction</summary>

```python
from argparse import ArgumentParser
import torch

class AddArgparseArgsExampleInvalidClass:
    def  __init__(self, invalid_class: torch.utils.data.Dataset):
        pass

def test_invalid_add_argparse_args():
    parser = ArgumentParser()
    add_argparse_args(
        AddArgparseArgsExampleInvalidClass,
        parser,
    )
```

</details>

<details>
<summary>stack traces</summary>

```
__________________________________________________________________________________ test_invalid_add_argparse_args __________________________________________________________________________________

    def test_invalid_add_argparse_args():
        parser = ArgumentParser()
        add_argparse_args(
            AddArgparseArgsExampleInvalidClass,
>           parser,
        )

tests/utilities/test_argparse.py:190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytorch_lightning/utilities/argparse.py:217: in add_argparse_args
    args_and_types = get_init_arguments_and_types(symbol)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'tests.utilities.test_argparse.AddArgparseArgsExampleInvalidClass'>

    def get_init_arguments_and_types(cls: Any) -> List[Tuple[str, Tuple, Any]]:
        r"""Scans the class signature and returns argument names, types and default values.
    
        Returns:
            List with tuples of 3 values:
            (argument name, set with argument types, argument default value).
    
        Examples:
    
            >>> from pytorch_lightning import Trainer
            >>> args = get_init_arguments_and_types(Trainer)
    
        """
        cls_default_params = inspect.signature(cls).parameters
        name_type_default = []
        for arg in cls_default_params:
            arg_type = cls_default_params[arg].annotation
            arg_default = cls_default_params[arg].default
            try:
>               arg_types = tuple(arg_type.__args__)
E               TypeError: 'NoneType' object is not iterable

pytorch_lightning/utilities/argparse.py:141: TypeError
```

</details>

Related: https://github.com/hsolbrig/PyShEx/issues/17#issuecomment-427557545

akihironitta commented:
I just submitted an issue to PL. Let's see how that behaviour gets handled in the linked issue for now :)

chris-clem commented:
> I just submitted an issue to PL. Let's see how that behaviour gets handled in the linked issue for now :)

Great, thank you!

chris-clem commented:
@akihironitta The bug seems to be fixed and merge. Is there a way to trigger the actions again?

akihironitta commented:
Since the fix has been merged and published in 1.4.8, I'm updating the min version of PL.

akihironitta commented:
Okay, there are quite a number of tests failing, so let me have a look to make sure that it's not caused by the changes in this PR...

akihironitta reviewed (approved):
@chris-clem Hi, thanks for reminding me of this issue. I had a look at all the failing tests, and am sure that they are all irrelevant to the change in this PR :]

chris-clem commented:
@Borda @akihironitta Should I fix the CodeFactor and DeepSource issues?

Borda commented:
> @Borda @akihironitta Should I fix the CodeFactor and DeepSource issues?

I think it is fine :] just adding back mypy typing... #776
